### ISSUE TITLE
Plugin GUI

### ISSUE THREAD

**Issue by Qualphey** [wmedrano/livi-rs#36]
Is there any way this library could be used to run graphical user interfaces of plugins? AFAIK lv2 plugins have their own custom GUI implementations independent of the specification. I'm looking for ways to identify these and run them. I've noticed `lilv_plugin_get_uis()` in the C library which is supposed to retrieve UI data.

**Comment by wmedrano** [comment]
I personally haven't used it. Did more digging and I think I found a C++ program that uses UIs from `lilv`: https://sourcegraph.com/github.com/zrythm/zrythm@3b96d59f430c0a46050581473c9675807435095b/-/blob/src/plugins/lv2/lv2_ui.c?L395

I can probably look into it sometime in February or March but let me know if you are interested in implementing it yourself and submitting a PR.

**Comment by Qualphey** [comment]
Yeah I'd be interested. You could give me some bullet points on relevant stuff I need to know.

Do you have any communication channel for contributors?

**Comment by wmedrano** [comment]
Probably the high level plan should be:

1. Create an example that uses the raw `lilv` API. If you want to still use the `livi` layer as a helper, you can call [`Plugin::raw`](https://docs.rs/livi/latest/livi/struct.Plugin.html). This object is very similar to the `C` object.
2. Fiddle around with the C api until it works :shrug: . Sorry, couldn't find much besides [`zrythm`](https://sourcegraph.com/github.com/zrythm/zrythm@3b96d59f430c0a46050581473c9675807435095b/-/blob/src/plugins/lv2/lv2_ui.c?L395) usage of it.

At some point, you probably want to call something like [`lilv_ui_get_binary_uri`](https://drobilla.net/docs/lilv/index.html#c.lilv_ui_get_binary_uri) to probably load a dynamic library. Here's an example where someone uses `libloading` to open a shared library, load a function from it, and run said function: https://michael-f-bryan.github.io/rust-ffi-guide/dynamic_loading.html#loading-the-contrived-example
